dev/wddr/device.c

```diff
@@ -531,38 +528,60 @@ wddr_return_t wddr_send_packets(wddr_handle_t wddr, const List_t *packets)
 }
 
 __attribute__(( weak ))
-wddr_return_t wddr_validate_recv_data(wddr_handle_t wddr,
-                                      const command_data_t *data,
-                                      burst_length_t bl,
-                                      bool *valid)
+bool wddr_validate_recv_data(wddr_handle_t wddr,
+                             const command_data_t *data,
+                             burst_length_t bl,
+                             uint8_t dq_byte_mask)
 {
-    uint8_t tmp, freq_id, ratio;
-    uint8_t byte = WDDR_DQ_BYTE_0;
+    uint8_t packet_data[PACKET_MAX_NUM_PHASES];
+    dfi_rx_packet_t packet;
+    uint8_t byte_data_idx[WDDR_DQ_BYTE_TOTAL] = {0};
+    uint8_t done = dq_byte_mask & ((1 << WDDR_DQ_BYTE_TOTAL) - 1);
+    uint8_t phases = 2 << wddr->dram.cfg->ratio;
+    uint8_t num_packets = bl >> wddr->dram.cfg->ratio >> 1;
+
+    // NOTE: Not validating that data is consecutive, but it should be
+    do
+    {
+        // Read packet
+        if (dfi_buffer_read_packet(&wddr->dfi, &packet) != DFI_SUCCESS)
+        {
+            return false;
+        }
 
-    // Initialize
-    pll_get_current_freq(&wddr->pll, &freq_id);
-    ratio = wddr->table->cfg.freq[freq_id].dram.ratio;
+        // Compare data for each byte
+        for (uint8_t byte = WDDR_DQ_BYTE_0; byte < WDDR_DQ_BYTE_TOTAL; byte++)
+        {
+            uint8_t byte_mask = 1 << byte;
+            // Don't process if byte_mask isn't set
+            if (!(done & byte_mask))
+            {
+                continue;
+            }
 
-    uint8_t num_packets = bl >> ratio >> 1;
-    if (dfi_buffer_read_packets(&wddr->dfi, &rx_buffer, num_packets) == DFI_ERROR_FIFO_EMPTY)
-    {
-        return WDDR_ERROR_FIFO_EMPTY;
+            // If not valid, then ignore
+            if (!extract_packet_data(&packet.packet, packet_data, byte, phases))
+            {
+                continue;
+            }
+
+            // Compare data
+            uint8_t idx = byte_data_idx[byte]++;
+            if (!compare_packet_data(packet_data, &data->dq[byte][idx * phases], phases, PACKET_DATA_MASK_BOTH))
+            {
+                return false;
+            }
+
+            // Clear mask to indicate this is done
+            if (idx == num_packets - 1)
+            {
+                done &= ~byte_mask;
+            }
+        }
     }
+    while(done); // Either no more packets to read or no more bytes to process
```

Comment:
Shouldn't this read as "while(!done);"

Reply:
It's correct. It's the variable name that's misleading. I will change that.

Reply:
The variable represents not_done, which is why you were trying to add the '!'. But changing that would change the logic. instead, I change it to remaining_byte_mask. So now it reads `while(remaining_byte_mask)` meaning that it will continue to loop until remaining_byte_mask is 0 which means the loop is done.

Reply:
Perfect, I think this reads much better